File: harness/src/livingdict/host.py
"""Shared capability host: the only I/O a Living Dictionary arm may perform."""

from __future__ import annotations

import hashlib
from collections.abc import Callable
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from .check import detect_check, run_workspace_check
from .gates import run_gates as measure_gates
from .policy import PathPolicy, changed_files, snapshot, workspace_digest
from .receipts import write_receipt
from .trace import emit, make_event
from .wave import METRIC_KEYS


class CapabilityError(Exception):
    """Typed trap. `code` is stable; arms and traces should key off it."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code
        self.message = message

    def as_dict(self) -> dict[str, str]:
        return {"code": self.code, "message": self.message}


@dataclass
class CapabilityHost:
    workspace: Path
    allowed_effects: tuple[str, ...]
    allowed_globs: tuple[str, ...]
    forbidden_globs: tuple[str, ...]
    trace_path: Path | None = None
    receipt_path: Path | None = None
    run_id: str = ""
    task_id: str = ""
    test_timeout_seconds: int = 60
    node_start_hook: Callable[[str], None] | None = None
    _policy: PathPolicy = field(init=False, repr=False)
    _before: dict[str, str] = field(init=False, repr=False)
    _effects_used: set[str] = field(init=False, repr=False)

    def __post_init__(self) -> None:
        self.workspace = Path(self.workspace).resolve()
        if not self.workspace.is_dir():
            raise CapabilityError("workspace", f"workspace is not a directory: {self.workspace}")
        self.allowed_effects = tuple(self.allowed_effects)
        self.allowed_globs = tuple(self.allowed_globs)
        self.forbidden_globs = tuple(self.forbidden_globs)
        if self.trace_path is not None:
            self.trace_path = Path(self.trace_path)
        if self.receipt_path is not None:
            self.receipt_path = Path(self.receipt_path)
        self._policy = PathPolicy(self.workspace, self.allowed_globs, self.forbidden_globs)
        self._before = snapshot(self.workspace)
        self._effects_used = set()
        self._last_check: dict[str, Any] | None = None
        self._event_sink: list[dict[str, Any]] | None = None
        self._write_receipt = True
        self._outer_policy: PathPolicy | None = None
        self.graph_metrics: dict[str, Any] = {}

    @classmethod
    def from_request(cls, request: dict[str, Any]) -> CapabilityHost:
        task = request["task"]
        return cls(
            workspace=Path(request["workspace"]),
            allowed_effects=tuple(task.get("allowed_effects", ("read", "write", "exec"))),
            allowed_globs=tuple(task["allowed_globs"]),
            forbidden_globs=tuple(task.get("forbidden_globs", ())),
            trace_path=Path(request["trace_path"]),
            receipt_path=Path(request["receipt_path"]),
            run_id=str(request.get("run_id", "")),
            task_id=str(task.get("id", "")),
        )

    def node_view(
        self,
        write_globs: tuple[str, ...] | list[str],
        extra_forbidden: tuple[str, ...] | list[str] = (),
    ) -> CapabilityHost:
        """Isolated host for one graph node. Shares the workspace, not mutable run state.

        `extra_forbidden` is sibling write-sets in the same wave so a doctored
        overlapping plan is still refused by PathPolicy (no partial mutation).
        """
        view = object.__new__(CapabilityHost)
        view.workspace = self.workspace
        view.allowed_effects = self.allowed_effects
        view.allowed_globs = tuple(write_globs)
        view.forbidden_globs = tuple(self.forbidden_globs) + tuple(extra_forbidden)
        view.trace_path = self.trace_path
        view.receipt_path = self.receipt_path
        view.run_id = self.run_id
        view.task_id = self.task_id
        view.test_timeout_seconds = self.test_timeout_seconds
        view.node_start_hook = None
        view._policy = PathPolicy(self.workspace, view.allowed_globs, view.forbidden_globs)
        view._before = self._before
        view._effects_used = set()
        view._last_check = None
        view._event_sink = []
        view._write_receipt = False
        view._outer_policy = self._outer_policy or self._policy
        view.graph_metrics = {}
        return view

    def emit_event(self, event_type: str, data: dict[str, Any] | None = None) -> None:
        if self._event_sink is not None:
            self._event_sink.append(make_event(event_type, data))
            return
        emit(self.trace_path, event_type, data)

    def absorb(self, view: CapabilityHost) -> None:
        self._effects_used.update(view._effects_used)
        if view._last_check is not None:
            self._last_check = view._last_check

    # --- words -----------------------------------------------------------

    def read_file(self, path: str) -> str:
        self._require_effect("read")
        target = self._existing_file(path)
        self._tool("READ-FILE", {"path": self._rel(path)})
        try:
            return target.read_text(encoding="utf-8")
        except UnicodeDecodeError as exc:
            raise CapabilityError("decode", f"not utf-8 text: {self._rel(path)}") from exc

    def list_dir(self, path: str = ".") -> list[str]:
        self._require_effect("read")
        target = self._existing_dir(path)
        self._tool("LIST-DIR", {"path": self._rel(path) or "."})
        names: list[str] = []
        for child in sorted(target.iterdir(), key=lambda item: item.name):
            rel = child.relative_to(self.workspace).as_posix()
            names.append(rel + ("/" if child.is_dir() else ""))
        return names

    def search(self, query: str) -> list[dict[str, Any]]:
        self._require_effect("read")
        self._tool("SEARCH", {"query": query})
        hits: list[dict[str, Any]] = []
        if query == "":
            return hits
        for file_path in sorted(self.workspace.rglob("*")):
            if not file_path.is_file() or any(
                part in {".git", "__pycache__"} for part in file_path.parts
            ):
                continue
            try:
                text = file_path.read_text(encoding="utf-8")
            except (UnicodeDecodeError, OSError):
                continue
            rel = file_path.relative_to(self.workspace).as_posix()
            for number, line in enumerate(text.splitlines(), start=1):
                if query in line:
                    hits.append({"path": rel, "line": number, "text": line})
        return hits

    def write_file(self, content: str, path: str) -> dict[str, Any]:
        self._require_effect("write")
        rel = self._rel(path)
        reason = self._policy.write_allowed(rel)
        if reason is None and self._outer_policy is not None:
            reason = self._outer_policy.write_allowed(rel)
        if reason is not None:
            self._tool("WRITE-FILE", {"path": rel, "denied": True})
            self.emit_event("execution.trap", {"reason": "policy", "detail": reason, "path": rel})
            raise CapabilityError("policy", reason)
        target = self._policy.resolve(path)
        data = content.encode("utf-8")
        digest = hashlib.sha256(data).hexdigest()
        receipt = {"path": rel, "bytes": len(data), "sha256": digest}
        if target.is_file() and target.read_bytes() == data:
            self._tool("WRITE-FILE", {"path": rel, "bytes": len(data), "idempotent": True})
            return receipt
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_bytes(data)
        self._tool("WRITE-FILE", {"path": rel, "bytes": len(data)})
        self.emit_event("mutation.applied", {"path": rel, "sha256": digest})
        return receipt

    def run_gates(self, word: str = "RUN-GATES", persist: bool = True) -> dict[str, Any]:
        self._require_effect("exec")
        spec = detect_check(self.workspace)
        self._tool(word, {"command": spec.get("label") or "RUN-GATES"})
        receipt = measure_gates(self.workspace, float(self.test_timeout_seconds), persist=persist)
        self._last_check = receipt
        if receipt.get("timed_out"):
            self.emit_event("execution.trap", {"reason": "test_timeout"})
        return receipt

    def run_tests(self) -> dict[str, Any]:
        """Product check only. Does not write .sb discharge into the workspace."""
        self._require_effect("exec")
        spec = detect_check(self.workspace)
        self._tool("RUN-TESTS", {"command": spec.get("label") or "RUN-TESTS"})
        receipt = run_workspace_check(self.workspace, float(self.test_timeout_seconds))
        self._last_check = receipt
        if receipt.get("timed_out"):
            self.emit_event("execution.trap", {"reason": "test_timeout"})
        return receipt

    def receipt(self, extra: dict[str, Any] | None = None) -> dict[str, Any]:
        after = snapshot(self.workspace)
        changed = changed_files(self._before, after)
        violations = [item for item in changed if self._policy.write_allowed(item) is not None]
        payload: dict[str, Any] = {
            "run_id": self.run_id,
            "task_id": self.task_id,
            "success": not violations,
            "workspace_before": workspace_digest(self._before),
            "workspace_after": workspace_digest(after),
            "changed_files": changed,
            "effects_used": sorted(self._effects_used),
            "policy_violations": [f"path outside policy after the fact: {item}" for item in violations],
        }
        if self._last_check is not None:
            payload["check"] = self._last_check
        for key in METRIC_KEYS:
            if key in self.graph_metrics and key not in payload:
                payload[key] = self.graph_metrics[key]
        if extra:
            payload.update(extra)
        if not self._write_receipt:
            self._tool("RECEIPT", {"path": None, "changed_files": changed})
            return payload
        target = self.receipt_path
        if target is None:
            target = self.workspace / "receipt.json"
        body = write_receipt(target, payload)
        self._tool("RECEIPT", {"path": str(target), "changed_files": changed})
        return body

    # --- internals -------------------------------------------------------

    def _require_effect(self, effect: str) -> None:
        if effect not in self.allowed_effects:
            self.emit_event("execution.trap", {"reason": "effect", "effect": effect})
            raise CapabilityError("effect", f"effect not allowed: {effect}")
        self._effects_used.add(effect)

    def _rel(self, path: str) -> str:
        try:
            return self._policy.relative(path)
        except ValueError as exc:
            self.emit_event("execution.trap", {"reason": "path", "detail": str(exc)})
            raise CapabilityError("path", str(exc)) from exc

    def _existing_file(self, path: str) -> Path:
        rel = self._rel(path)
        target = self.workspace / rel if rel else self.workspace
        if not target.is_file():
            self.emit_event("execution.trap", {"reason": "missing_file", "path": rel or "."})
            raise CapabilityError("missing_file", f"missing file: {rel or '.'}")
        return target

    def _existing_dir(self, path: str) -> Path:
        if path in {"", "."}:
            return self.workspace
        rel = self._rel(path)
        target = self.workspace / rel if rel else self.workspace
        if not target.is_dir():
            self.emit_event("execution.trap", {"reason": "missing_file", "path": rel or "."})
            raise CapabilityError("missing_file", f"missing directory: {rel or '.'}")
        return target

    def _tool(self, name: str, data: dict[str, Any]) -> None:
        payload = {"tool": name, **data}
        self.emit_event("tool.call", payload)

File: beam/lib/ld_host/host.ex
defmodule LdHost.Host do
  @moduledoc """
  The capability host: the only I/O a Living Dictionary arm may perform.

  Port of `harness/src/livingdict/host.py` — same words, same trap codes,
  same receipt shape. Events go through the `emit` closure (wired to the
  run's Ledger); the host itself owns no process state, it is a value
  threaded through the Forth VM.

  Documented divergence from the frozen Python reference: READ-FILE on a
  non-UTF-8 file does NOT trap `decode`. It returns a summary value string
  (`"<<binary file: N bytes, magic xxxxxxxx 'printable head'>>"`) so the
  planner can observe the file exists and is binary instead of retrying
  the read and burning episodes. The `decode` trap code itself remains
  reserved for any other path that used it.
  """

  @behaviour LdHost.Capability

  alias LdHost.{Policy, Cmd}

  defstruct workspace: nil,
            allowed_effects: ["read", "write", "exec"],
            allowed_globs: [],
            forbidden_globs: [],
            run_id: "",
            task_id: "",
            episode: 0,
            test_timeout_seconds: 60,
            emit: nil,
            policy: nil,
            outer_policy: nil,
            before: %{},
            effects_used: MapSet.new(),
            last_check: nil,
            contract: nil,
            allow_model_checks: false,
            objects_dir: nil,
            receipt_path: nil,
            write_receipt?: true

  def new(workspace, opts \\ []) do
    workspace = Path.expand(workspace)

    unless File.dir?(workspace) do
      raise ArgumentError, "workspace is not a directory: #{workspace}"
    end

    allowed = Keyword.get(opts, :allowed_globs, ["**"])
    forbidden = Keyword.get(opts, :forbidden_globs, [])

    %__MODULE__{
      workspace: workspace,
      allowed_effects: Keyword.get(opts, :allowed_effects, ["read", "write", "exec"]),
      allowed_globs: allowed,
      forbidden_globs: forbidden,
      run_id: Keyword.get(opts, :run_id, ""),
      task_id: Keyword.get(opts, :task_id, ""),
      episode: Keyword.get(opts, :episode, 0),
      test_timeout_seconds: Keyword.get(opts, :test_timeout_seconds, 60),
      emit: Keyword.get(opts, :emit, fn _type, _data -> :ok end),
      policy: Policy.new(workspace, allowed, forbidden),
      before: Policy.snapshot(workspace),
      contract: Keyword.get(opts, :contract),
      allow_model_checks: Keyword.get(opts, :allow_model_checks, false),
      objects_dir: Keyword.get(opts, :objects_dir),
      receipt_path: Keyword.get(opts, :receipt_path),
      write_receipt?: Keyword.get(opts, :write_receipt?, true)
    }
  end

  @doc """
  Isolated host for one graph node: narrowed write globs, sibling write
  sets forbidden, buffered-elsewhere event emission, no receipt writing.
  """
  def node_view(%__MODULE__{} = host, write_globs, extra_forbidden, emit) do
    forbidden = host.forbidden_globs ++ extra_forbidden

    %{
      host
      | allowed_globs: write_globs,
        forbidden_globs: forbidden,
        policy: Policy.new(host.workspace, write_globs, forbidden),
        outer_policy: host.outer_policy || host.policy,
        effects_used: MapSet.new(),
        last_check: nil,
        emit: emit,
        write_receipt?: false
    }
  end

  @doc "Merge a node view's observations back into the parent host."
  def absorb(%__MODULE__{} = host, %__MODULE__{} = view) do
    %{
      host
      | effects_used: MapSet.union(host.effects_used, view.effects_used),
        last_check: view.last_check || host.last_check
    }
  end

  # ---- Capability dispatch ---------------------------------------------

  @impl true
  def call(host, "READ-FILE", [path]), do: read_file(host, path)
  def call(host, "LIST-DIR", [path]), do: list_dir(host, path)
  def call(host, "SEARCH", [query]), do: search(host, query)
  def call(host, "WRITE-FILE", [content, path]), do: write_file(host, content, path)
  def call(host, "USE-OBJECT", [sha]), do: fetch_object(host, sha)
  def call(host, "PATCH-FILE", [patch, path]), do: patch_file(host, patch, path)
  def call(host, "RUN-TESTS", []), do: run_checks(host, "RUN-TESTS", false)
  def call(host, "RUN-GATES", []), do: run_checks(host, "RUN-GATES", true)
  def call(host, "RECEIPT", []), do: receipt(host)

  # ---- words ------------------------------------------------------------

  def read_file(host, path) do
    with {:ok, host} <- require_effect(host, "read"),
         {:ok, rel} <- rel(host, path),
         target = target_path(host, rel),
         true <- File.regular?(target) || {:missing, rel} do
      tool(host, "READ-FILE", %{path: rel})

      case File.read(target) do
        {:ok, data} ->
          if String.valid?(data) do
            {:ok, data, host}
          else
            # Divergence from the Python reference (which traps "decode"):
            # a binary read returns a summary so the planner does not
            # blindly retry the read. See the moduledoc.
            {:ok, binary_summary(data), host}
          end

        {:error, _} ->
          trap_event(host, %{reason: "missing_file", path: empty_dot(rel)})
          {:trap, "missing_file", "missing file: #{empty_dot(rel)}"}
      end
    else
      {:missing, rel} ->
        trap_event(host, %{reason: "missing_file", path: empty_dot(rel)})
        {:trap, "missing_file", "missing file: #{empty_dot(rel)}"}

      {:trap, _, _} = trap ->
        trap
    end
  end

  def list_dir(host, path) do
    with {:ok, host} <- require_effect(host, "read"),
         {:ok, rel} <- rel_dir(host, path),
         target = target_path(host, rel),
         true <- File.dir?(target) || {:missing, rel} do
      tool(host, "LIST-DIR", %{path: empty_dot(rel)})

      names =
        target
        |> File.ls!()
        |> Enum.sort()
        |> Enum.map(fn name ->
          rel_child = if rel == "", do: name, else: "#{rel}/#{name}"
          if File.dir?(Path.join(target, name)), do: rel_child <> "/", else: rel_child
        end)

      {:ok, names, host}
    else
      {:missing, rel} ->
        trap_event(host, %{reason: "missing_file", path: empty_dot(rel)})
        {:trap, "missing_file", "missing directory: #{empty_dot(rel)}"}

      {:trap, _, _} = trap ->
        trap
    end
  end

  def search(host, query) do
    with {:ok, host} <- require_effect(host, "read") do
      tool(host, "SEARCH", %{query: query})

      hits =
        if query == "" do
          []
        else
          host.workspace
          |> Policy.snapshot()
          |> Map.keys()
          |> Enum.sort()
          |> Enum.flat_map(fn rel ->
            case File.read(Path.join(host.workspace, rel)) do
              {:ok, text} when is_binary(text) ->
                if String.valid?(text) do
                  text
                  |> String.split("\n")
                  |> Enum.with_index(1)
                  |> Enum.filter(fn {line, _} -> String.contains?(line, query) end)
                  |> Enum.map(fn {line, number} -> %{path: rel, line: number, text: line} end)
                else
                  []
                end

              _ ->
                []
            end
          end)
        end

      {:ok, hits, host}
    end
  end

  def write_file(host, content, path) do
    with {:ok, host} <- require_effect(host, "write"),
         {:ok, rel} <- rel(host, path) do
      reason =
        Policy.write_allowed(host.policy, rel) ||
          (host.outer_policy && Policy.write_allowed(host.outer_policy, rel))

      if reason do
        tool(host, "WRITE-FILE", %{path: rel, denied: true})
        trap_event(host, %{reason: "policy", detail: reason, path: rel})
        {:trap, "policy", reason}
      else
        target = Path.join(host.workspace, rel)
        digest = Policy.sha256_hex(content)
        intern_blob(host, content, digest)
        receipt = %{path: rel, bytes: byte_size(content), sha256: digest}

        case File.read(target) do
          {:ok, existing} when existing == content ->
            tool(host, "WRITE-FILE", %{path: rel, bytes: byte_size(content), idempotent: true})
            {:ok, receipt, host}

          _ ->
            File.mkdir_p!(Path.dirname(target))
            File.write!(target, content)
            tool(host, "WRITE-FILE", %{path: rel, bytes: byte_size(content)})
            emit(host, "mutation.applied", %{path: rel, sha256: digest})
            {:ok, receipt, host}
        end
      end
    end
  end

  defp run_checks(host, word, persist?) do
    with {:ok, host} <- require_effect(host, "exec") do
      tool(host, word, %{command: word})
      report = LdHost.Gates.run(host, persist?: persist?)
      host = %{host | last_check: report}

      if report[:timed_out] do
        trap_event(host, %{reason: "test_timeout"})
      end

      {:ok, report, host}
    end
  end

  def receipt(host, extra \\ %{}) do
    after_snap = Policy.snapshot(host.workspace)
    changed = Policy.changed_files(host.before, after_snap)

    violations =
      Enum.filter(changed, fn rel -> Policy.write_allowed(host.policy, rel) != nil end)

    payload =
      %{
        run_id: host.run_id,
        task_id: host.task_id,
        success: violations == [],
        workspace_before: Policy.workspace_digest(host.before),
        workspace_after: Policy.workspace_digest(after_snap),
        changed_files: changed,
        effects_used: host.effects_used |> MapSet.to_list() |> Enum.sort(),
        policy_violations:
          Enum.map(violations, &"path outside policy after the fact: #{&1}")
      }
      |> maybe_put(:check, host.last_check)
      |> Map.merge(extra)

    if host.write_receipt? do
      target = host.receipt_path || Path.join(host.workspace, "receipt.json")
      File.write!(target, JSON.encode!(payload))
      tool(host, "RECEIPT", %{path: target, changed_files: changed})
    else
      tool(host, "RECEIPT", %{path: nil, changed_files: changed})
    end

    {:ok, payload, host}
  end

  @doc "CAS intern. Identical bytes reuse the same objects/<aa>/<sha> path."
  def intern(%__MODULE__{} = host, content) when is_binary(content) do
    digest = Policy.sha256_hex(content)
    intern_blob(host, content, digest)
    digest
  end

  @doc "Read an interned blob. Requires the read effect (USE-OBJECT)."
  def fetch_object(host, sha) do
    with {:ok, host} <- require_effect(host, "read") do
      tool(host, "USE-OBJECT", %{sha256: sha})

      case read_blob(host.objects_dir, sha) do
        {:ok, data} ->
          {:ok, data, host}

        {:error, reason} ->
          trap_event(host, %{reason: "missing_object", sha256: sha, detail: reason})
          {:trap, "missing_object", "no object: #{sha}"}
      end
    end
  end

  def patch_file(host, patch, path) do
    with {:ok, current, host} <- read_file(host, path) do
      write_file(host, apply_patch(current, patch), path)
    end
  end

  # ---- internals --------------------------------------------------------

  defp require_effect(host, effect) do
    if effect in host.allowed_effects do
      {:ok, %{host | effects_used: MapSet.put(host.effects_used, effect)}}
    else
      trap_event(host, %{reason: "effect", effect: effect})
      {:trap, "effect", "effect not allowed: #{effect}"}
    end
  end

  defp rel(host, path) do
    case Policy.relative(host.policy, path) do
      {:ok, rel} ->
        {:ok, rel}

      {:error, reason} ->
        trap_event(host, %{reason: "path", detail: reason})
        {:trap, "path", reason}
    end
  end

  defp rel_dir(_host, path) when path in ["", "."], do: {:ok, ""}
  defp rel_dir(host, path), do: rel(host, path)

  defp target_path(host, ""), do: host.workspace
  defp target_path(host, rel), do: Path.join(host.workspace, rel)

  defp empty_dot(""), do: "."
  defp empty_dot(rel), do: rel

  def intern_blob(%{objects_dir: nil}, _content, _digest), do: :ok

  def intern_blob(%{objects_dir: dir}, content, digest) when is_binary(dir) and is_binary(digest) do
    path = blob_path(dir, digest)

    unless path == nil or File.exists?(path) do
      File.mkdir_p!(Path.dirname(path))
      File.write!(path, content)
    end

    :ok
  end

  def intern_blob(_host, _content, _digest), do: :ok

  defp read_blob(nil, _sha), do: {:error, "no objects_dir"}

  defp read_blob(dir, sha) do
    case blob_path(dir, sha) do
      nil ->
        {:error, "invalid sha"}

      path ->
        case File.read(path) do
          {:ok, data} -> {:ok, data}
          {:error, _} -> {:error, "missing"}
        end
    end
  end

  defp blob_path(dir, sha) when is_binary(sha) do
    if sha =~ ~r/^[0-9a-f]{64}$/ do
      Path.join([dir, String.slice(sha, 0, 2), sha])
    end
  end

  defp blob_path(_, _), do: nil

  # First-occurrence find/replace when the patch has a >>> separator;
  # otherwise the patch is the whole new body (file must already exist).
  defp apply_patch(current, patch) do
    case String.split(patch, ">>>", parts: 2) do
      [find, replace] -> String.replace(current, find, replace, global: false)
      _ -> patch
    end
  end

  # Summary value for a non-UTF-8 read: size, first 4 bytes as hex, and
  # up to 16 leading printable chars (e.g. a magic string like
  # "SQLite format 3").
  defp binary_summary(data) do
    size = byte_size(data)
    magic = data |> binary_part(0, min(4, size)) |> Base.encode16(case: :lower)

    head =
      data
      |> :binary.bin_to_list(0, min(16, size))
      |> Enum.take_while(&(&1 in 32..126))
      |> List.to_string()

    if head == "" do
      "<<binary file: #{size} bytes, magic #{magic}>>"
    else
      "<<binary file: #{size} bytes, magic #{magic} '#{head}'>>"
    end
  end

  defp tool(host, name, data), do: emit(host, "tool.call", Map.put(data, :tool, name))
  defp trap_event(host, data), do: emit(host, "execution.trap", data)
  defp emit(%{emit: emit}, type, data), do: emit.(type, data)

  defp maybe_put(map, _key, nil), do: map
  defp maybe_put(map, key, value), do: Map.put(map, key, value)

  @doc "Timeout in ms for one gate command, from the host default."
  def timeout_ms(%__MODULE__{test_timeout_seconds: s}), do: s * 1000

  def cmd(host, command, timeout_ms), do: Cmd.sh(command, host.workspace, timeout_ms)
end
